service/frontend/nexus_incoming_service_client.go

```diff
@@ -0,0 +1,373 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package frontend
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/google/uuid"
+	"go.temporal.io/api/nexus/v1"
+	"go.temporal.io/api/operatorservice/v1"
+	"go.temporal.io/api/serviceerror"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"go.temporal.io/server/api/matchingservice/v1"
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+	hlc "go.temporal.io/server/common/clock/hybrid_logical_clock"
+	"go.temporal.io/server/common/dynamicconfig"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/namespace"
+	cnexus "go.temporal.io/server/common/nexus"
+	p "go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/rpc"
+)
+
+const (
+	reservedTaskQueuePrefix = "/_sys/"
+)
+
+// ServiceNameRegex is the regular expression that incoming service names must match.
+var ServiceNameRegex = regexp.MustCompile(`[a-zA-Z_][a-zA-Z0-9_]*`)
+
+type (
+	// NexusIncomingServiceClient manages frontend CRUD requests for Nexus incoming services.
+	// Create, Update, and Delete requests are forwarded to matching service which owns the incoming services table.
+	// Read (Get and List) requests are sent directly to persistence. This is to ensure read-after-write consistency.
+	NexusIncomingServiceClient struct {
+		config *nexusIncomingServiceClientConfig
+
+		namespaceRegistry namespace.Registry // used to validate referenced namespaces exist
+		matchingClient    matchingservice.MatchingServiceClient
+		persistence       p.NexusIncomingServiceManager
+
+		logger log.Logger
+	}
+
+	nexusIncomingServiceClientConfig struct {
+		maxNameLength       dynamicconfig.IntPropertyFn
+		maxTaskQueueLength  dynamicconfig.IntPropertyFn
+		maxSize             dynamicconfig.IntPropertyFn
+		listDefaultPageSize dynamicconfig.IntPropertyFn
+		listMaxPageSize     dynamicconfig.IntPropertyFn
+	}
+)
+
+func newNexusIncomingServiceClientConfig(dc *dynamicconfig.Collection) *nexusIncomingServiceClientConfig {
+	return &nexusIncomingServiceClientConfig{
+		maxNameLength:       dc.GetIntProperty(dynamicconfig.NexusIncomingServiceNameMaxLength, 200),
+		maxTaskQueueLength:  dc.GetIntProperty(dynamicconfig.MaxIDLengthLimit, 1000),
```

Comment:
You shouldn't redefine this config and the default value, it's already here: https://github.com/temporalio/temporal/blob/a98b60e3f43f69b879f3fdab215817f39b4f5799/service/frontend/service.go#L243

I'm also a bit concerned about not sharing task queue name validation logic with workflow handler, let's maybe refactor the validation into a shared helper:
https://github.com/temporalio/temporal/blob/a98b60e3f43f69b879f3fdab215817f39b4f5799/service/frontend/workflow_handler.go#L4008

Reply:
I moved the logic for validating the namespace name to the frontend `validators.go` file since that is all that is needed for incoming services. I also left a TODO there about refactoring the validation logic to make it consistent across the codebase because I don't want to take that on as part of this PR.

Reply:
SGTM. Maybe just track it in an issue.

Reply:
Created https://github.com/temporalio/temporal/issues/5628